Feature/569 implement filtering on the room table

Urjikoo opened the pull request:
## Types of changes
<!--
  What types of changes does your code introduce? Put an `x` in all the boxes
  that apply:
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality
      to change)

## Checklist:
<!--
  Go over all the following points, and put an `x` in all the boxes that apply.
-->
- [x] I have run `eslint` on the code
- [x] I have added JSDoc for all of my code (where applicable)
- [x] I have added tests to cover my changes.

## Priority:
- [x] Normal <!-- New piece of functionality -->
- [ ] High <!-- Critical bug requiring urgent review -->

## Related Issues:
<!--
  Add the issue number in below that this PR is intended to address.
[#569]
  Also mention any issues that this PR is related to, and if possible,
  provide more information regarding the relationship to the issues in
  question (i.e does the PR fix the issue, is it designed to fix another bug
  that is similar to the issue etc.)
-->
Fixes #569

<!--
  Attribution:
  PR Template adapted from: https://github.com/h5bp/html5-boilerplate/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

Urjikoo committed:
The table filter dropdowns and text input filds created.
Campus and building filter dropdowns and room and capacity input filters are created and function as intended.

Urjikoo committed:
Tests were added to check room admin table filters
Tests were added to check campus, building,room and capacity filters that were added on room admin table.

Urjikoo committed:
Tests were implemented removing unused variables

Urjikoo committed:
Implmented room admin test to account the filters row.

Urjikoo committed:
Capacity filter label re-added to resolve build fail

natalynyu commented on src/client/components/pages/RoomAdmin/__tests__/RoomAdminTable.test.tsx:
```diff
@@ -21,6 +24,186 @@ import {
+        );
+      });
+    });
+    it('filters table data when cambridge is selected', async function () {
```
Just a typo: capital C for Cambridge

natalynyu commented on src/client/components/pages/RoomAdmin/__tests__/RoomAdminTable.test.tsx:
```diff
@@ -21,6 +24,186 @@ import {
+    it('filters table data for selected value', async function () {
+      await wait(() => renderResult.getAllByRole('row').length > 1);
+      const [buildingDropDown] = renderResult.getAllByLabelText('The table will be filtered by selected building name');
+      fireEvent.change(buildingDropDown, { target: { value: 'Bauer Laboratory' } });
```
I think it would be helpful to future proof the code by replacing "Bauer Laboratory" with a reference to a room from `adminRoomsResponse` (e.g. `adminRoomsResponse[0].building.name`). This way, if the test data is changed in the future such that `Bauer Laboratory` is removed, the test will still work, because the building you are changing it to references what we are stubbing `getAdminRooms` to resolve to on line 32: 
`getStub.resolves(adminRoomsResponse);`

natalynyu commented on src/client/components/pages/RoomAdmin/__tests__/RoomAdminTable.test.tsx:
```diff
@@ -21,6 +24,186 @@ import {
+        within(row).queryAllByRole('columnheader').length === 0
+      ));
+      bodyRow.forEach((row) => {
+        strictEqual(
+          within(row).queryByText('Bauer Laboratory') !== null,
+          true,
+          'Bauer Laboratory is not in the table as expected'
+        );
```
...Then down here, if you define a variable above for `testBuilding` for instance, you can use template strings for the `strictEqual` call to replace `Bauer Laboratory`.

natalynyu commented on src/client/components/pages/RoomAdmin/__tests__/RoomAdminTable.test.tsx:
```diff
@@ -21,6 +24,186 @@ import {
+    const roomFilterInput = renderResult.getByLabelText('The table will be filtered by selected room name', { exact: true }) as HTMLSelectElement;
+    strictEqual(roomFilterInput.value, '');
+  });
+  it('filters table data when room name is entered', async function () {
+    await wait(() => renderResult.getAllByRole('row').length > 1);
+    const [roomInput] = renderResult.getAllByLabelText('The table will be filtered by selected room name');
+    fireEvent.change(roomInput, { target: { value: '101A' } });
+    await wait(() => renderResult.getAllByRole('row').length > 1);
+    const rows = renderResult.getAllByRole('row');
+    const bodyRow = rows.filter((row) => (
+      within(row).queryAllByRole('columnheader').length === 0
+    ));
+    bodyRow.forEach((row) => {
+      strictEqual(
+        within(row).queryByText('101A') !== null,
+        true,
+        '101A is not in the table as expected'
+      );
+    });
+  });
```
I would also recommend defining a `testRoomName` here that references `adminRoomsResponse` instead of hardcoding `101A`.

natalynyu commented on src/client/components/pages/RoomAdmin/__tests__/RoomAdminTable.test.tsx:
```diff
@@ -21,6 +24,186 @@ import {
+    const capacityFilterInput = renderResult.getByLabelText('The table will be filtered by entered capacity', { exact: true }) as HTMLSelectElement;
+    strictEqual(capacityFilterInput.value, '');
+  });
+  it('filters table data by entered capacity entered', async function () {
+    await wait(() => renderResult.getAllByRole('row').length > 1);
+    const [capacityFilterInput] = renderResult.getAllByLabelText('The table will be filtered by selected room name');
+    fireEvent.change(capacityFilterInput, { target: { value: '20' } });
+    await wait(() => renderResult.getAllByRole('row').length > 1);
+    const rows = renderResult.getAllByRole('row');
+    const bodyRow = rows.filter((row) => (
+      within(row).queryAllByRole('columnheader').length === 0
+    ));
+    bodyRow.forEach((row) => {
+      strictEqual(
+        within(row).queryByText('20') !== null,
+        true,
+        'room capacity filter is not populating data properly'
+      );
+    });
+  });
```
I would also define a `testCapacity` variable at the top of this test, since '20' is being referenced in more than one place. This is a good practice so that if you ever need to change the value from '20', you just have to make the variable a different value rather than changing the '20' value in multiple places. I know it's only being used twice here, but I think it's a good practice in general nonetheless.

natalynyu reviewed (changes requested):
This is looking great! I just had some notes mostly related to taking advantage of variables in tests and documentation.

Urjikoo committed:
building,room and capacity tests were updated.
implemented capacity filter to check if the table content is filtered by capacity.For future proofing, both building and room tests were updated with test data varialbe avoiding hard coded building and room names.

natalynyu commented on src/client/components/pages/RoomAdmin/RoomAdmin.tsx:
```diff
@@ -258,31 +251,34 @@ const RoomAdmin: FunctionComponent = (): ReactElement => {
                       <Dropdown
+                        hideError
+                        id="building-filter"
+                        label="Change to filter the list of meetings by building"
```
The room admin page is filtering rooms, not meetings, so this text should be updated.

natalynyu commented on src/client/components/pages/RoomAdmin/RoomAdmin.tsx:
```diff
@@ -292,11 +288,11 @@ const RoomAdmin: FunctionComponent = (): ReactElement => {
                         placeholder="Filter by Room Name"
                         value={roomFilter}
-                        label="The table will be filtered by selected room name"
+                        label="Change to filter the list of meetings by room"
```
The room admin is not related to meetings, so this label text should be updated as well.

natalynyu commented on src/client/components/pages/RoomAdmin/RoomAdmin.tsx:
```diff
@@ -308,11 +304,11 @@ const RoomAdmin: FunctionComponent = (): ReactElement => {
                         placeholder="Enter Capacity"
                         value={capacityFilter}
-                        label="The table will be filtered by entered capacity"
+                        label="Change to filter the list of meetings by capacity"
```
Same for this label - the room admin is not related to meetings, so this label text should be updated.

natalynyu reviewed (changes requested):
Thanks for taking a look at this, Urji - this is looking great! I left a few more comments mostly related to changing the text of the labels and formatting. I think that one of the comments from my previous review was missed regarding adding a docblock above the `filteredRoomList` function. In the future, it would be really helpful if there were smaller commits that grouped similar changes into each commit, rather than having one commit with all of the changes. It makes it easier to review what was changed. I found that the "Commit Guidelines" section of [these docs](https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project) Jonathan sent out last week on contributing to projects really helpful. Usually, I will try to comment back on each comment made on my PRs since this review process should be more like a conversation so that we can collaborate and find the best solution for issues we encounter. Let me know if you have any questions!

Urjikoo committed:
The spacing for buildingnamelist function comment after the initial asterisks are updated.

Urjikoo committed:
Label for room admin dropdown and textinput filters updated
Dropdown and text input labels were corrected from 'change to filter the list of meetings' to 'change to filter list of rooms'.

Urjikoo committed:
Roomadmin filter tests lables were updated and spacing on building dropdown test removed

Urjikoo committed:
Comment was added for filteredRoomList function.

natalynyu commented on src/client/components/pages/RoomAdmin/RoomAdmin.tsx:
```diff
@@ -292,7 +292,7 @@ const RoomAdmin: FunctionComponent = (): ReactElement => {
                         placeholder="Filter by Room Name"
                         value={roomFilter}
-                        label="Change to filter the list of meetings by room"
+                        label="Change to filter the list of rooms by room"
```
It probably makes more sense to have the text "Change to filter the list of rooms by room **name**" here, since the campus and building are also part of the room information in general.

natalynyu commented on src/client/components/pages/RoomAdmin/RoomAdmin.tsx:
```diff
@@ -138,7 +175,40 @@ const RoomAdmin: FunctionComponent = (): ReactElement => {
+    { value: buildingName, label: buildingName }
+  ));
+  /**
+   * Filters rooms list based on selected campus, building, room and capacity.
```
Small note, but I would also change "room" to "room name" here as well for clarity.

natalynyu commented on src/client/components/pages/RoomAdmin/RoomAdmin.tsx:
```diff
@@ -181,32 +251,105 @@ const RoomAdmin: FunctionComponent = (): ReactElement => {
+                      <TextInput
+                        id="capacity-filter"
+                        name="capacity-filter"
+                        placeholder="Enter Capacity"
```
For consistency with the other filter fields, I think the placeholder should be "Filter by Capacity".

natalynyu commented on src/client/components/pages/RoomAdmin/RoomAdmin.tsx:
```diff
@@ -138,7 +175,40 @@ const RoomAdmin: FunctionComponent = (): ReactElement => {
+  /**
+   * Populates building names for the building filter dropdown.
+   */
+  const buildingDropDown = buildingNameList.map((buildingName) => (
+    { value: buildingName, label: buildingName }
+  ));
```
I think that this list of buildings should be sorted alphabetically. Currently, the buildings are alphabetically sorted within each campus, but I think that it would be nice to sort all of them for a better user experience.
<img width="405" alt="image" src="https://user-images.githubusercontent.com/29589689/207969699-84a72a95-9f58-4631-8bcb-ecf185ba2a7e.png">

natalynyu reviewed (changes requested):
This is almost there! I just had a few comments mostly related to consistency.

Urjikoo committed:
Room filter label updated

Urjikoo committed:
The comment for filteredRoomList was updated

Urjikoo committed:
Capacity text input placeholder was updated

Urjikoo committed:
Building dropdown menu updated to be sorted in alphabetical order.

Urjikoo committed:
roomFilter input test updated with updated label name

natalynyu reviewed (changes requested):
Everything is working as expected, and it looks great. I just had 2 final recommendations related to types.

Urjikoo committed:
Updated type for campus and building dropdown filters.

natalynyu reviewed (approved):
LGTM! Approved.
